File: backend/src/Packets.kt
package me.schlaubi

import kotlinx.serialization.*
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.modules.SerializersModule
import kotlinx.serialization.modules.polymorphic
import java.util.*

internal val PacketModule = SerializersModule {
    polymorphic(Packet::class) {
        subclass(ClosePacket::class, ClosePacket.serializer())
        subclass(MessageCreatePacket::class, MessageCreatePacket.serializer())
        subclass(UserJoinedPacket::class, UserJoinedPacket.serializer())
        subclass(UserLeftPacket::class, UserLeftPacket.serializer())
        subclass(HelloPacket::class, HelloPacket.serializer())
    }
}

/**
 * Represents a packet sent through the websocket containing payload [data].
 */
@Serializable
abstract class Packet<T : Packet.Data> {
    /**
     * The payload of the packet.
     */
    @SerialName("d")
    abstract val data: T

    /**
     * Interface for payload.
     */
    interface Data

    /**
     * Object for empty payload
     */
    @Serializable(with = EmptyDataSerializer::class)
    object EmptyData : Data

    /**
     * Representation of supported opcodes.
     */
    @Serializable
    @Suppress("EnumEntryName")
    enum class OPCode {
        /**
         * Gracefully closes the connection. Contains [EmptyData].
         */
        @SerialName("close")
        CLOSE,

        /**
         * Event stating that a new message was sent. Contains [MessageCreatePacket.Data].
         */
        @SerialName("messageCreate")
        MESSAGE_CREATE,

        /**
         * Event stating that a new user joined the room. Contains [UserJoinedPacket.Data].
         */
        @SerialName("userJoined")
        USER_JOINED,

        /**
         * Event stating that a new user joined the room. Contains [UserLeftPacket.Data].
         */
        @SerialName("userLeft")
        USER_LEFT,

        /**
         * Event stating that the client successfully connected to the socket. Contains [HelloPacket.Data].
         */
        @SerialName("hello")
        HELLO
    }
}

/**
 * Sent by: client
 * OPCode: [Packet.OPCode.CLOSE]
 * Command sent to gracefully close the current connection.
 */
@SerialName("close")
@Serializable
class ClosePacket : Packet<Packet.EmptyData>() {
    override val data: EmptyData = EmptyData
}

/**
 * Sent by: server
 * OPCode: [Packet.OPCode.MESSAGE_CREATE]
 * Event stating a new message has been created.
 */
@SerialName("messageCreate")
@Serializable
class MessageCreatePacket(@SerialName("d") override val data: Data) :
    Packet<MessageCreatePacket.Data>() {

    /**
     * @see Packet.OPCode.MESSAGE_CREATE
     * @param content the content of the message
     * @param author the author of the message
     */
    @Serializable
    data class Data(val content: String, val author: String) : Packet.Data
}

/**
 * Sent by: server
 * OPCode: [Packet.OPCode.USER_JOINED]
 * Event stating a new user joined the channel
 */
@SerialName("userJoined")
@Serializable
class UserJoinedPacket(@SerialName("d") override val data: Data) : Packet<UserJoinedPacket.Data>() {

    /**
     * @see Packet.OPCode.USER_JOINED
     * @param user the [WebsocketUser] that joined
     */
    @Serializable
    data class Data(val user: WebsocketUser) : Packet.Data
}

/**
 * Sent by: server
 * OPCode: [Packet.OPCode.USER_LEFT]
 * Event stating a user left the channel
 */
@SerialName("userLeft")
@Serializable
class UserLeftPacket(@SerialName("d") override val data: Data) : Packet<UserLeftPacket.Data>() {

    /**
     * @see Packet.OPCode.USER_JOINED
     * @param user the [WebsocketUser] that left
     */
    @Serializable
    data class Data(val user: WebsocketUser) : Packet.Data
}

/**
 * Sent by: server
 * OPCode: [Packet.OPCode.HELLO]
 * Welcome event sent to client to update channel info
 */
@SerialName("hello")
@Serializable
class HelloPacket(@SerialName("d") override val data: Data) : Packet<HelloPacket.Data>() {

    /**
     * @see Packet.OPCode.HELLO
     * @property user the user connected
     * @property key they auth key for this user
     * @property users a list of users in this channel
     * @property channelId the id of the channel you connected to
     * @property channelName the name of the channel the user connected to
     */
    @Serializable
    data class Data(
        @Contextual val channelId: UUID,
        val user: WebsocketUser,
        val key: String,
        val channelName: String,
        val users: List<WebsocketUser>
    ) : Packet.Data
}

/**
 * A user with a [name] and [admin status](isAdmin) and a unique [nonce].
 *
 * @property name the name of the user
 * @property isAdmin whether the user created this channel or not
 * @property nonce a unique identifier for the user
 */
@Serializable
data class WebsocketUser(val name: String, val isAdmin: Boolean, val nonce: String)

internal object EmptyDataSerializer : KSerializer<Packet.EmptyData> {
    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("data", PrimitiveKind.STRING)

    override fun deserialize(decoder: Decoder): Packet.EmptyData = Packet.EmptyData

    @OptIn(ExperimentalSerializationApi::class)
    override fun serialize(encoder: Encoder, value: Packet.EmptyData): Unit = encoder.encodeNull()
}

File: client/lib/src/gateway/packets.dart
import 'package:json_annotation/json_annotation.dart';

part 'packets.g.dart';

/// Representation of Websocket packet
class WebsocketPacket<T> {
  /// The OPCode of the packet
  final OPCode code;

  /// The data payload
  final WebSocketData data;

  // For internal use only //

  WebsocketPacket(this.code, {this.data});

  Map<String, dynamic> toJson() => <String, dynamic>{
        'op': code.toString().substring(7),
        'd': data?.toJson() ?? <String, dynamic>{}
      };

  factory WebsocketPacket.fromJson(dynamic map) {
    var opCode = map['op'];
    var data = map['d'];

    if (opCode == 'messageCreate') {
      return WebsocketPacket(OPCode.messageCreate,
          data: MessageCreateData.fromJson(data));
    } else if (opCode == 'userJoined') {
      return WebsocketPacket(OPCode.userJoined,
          data: UserJoinedPacketData.fromMap(data));
    } else if (opCode == 'userLeft') {
      return WebsocketPacket(OPCode.userLeft,
          data: UserLeftPacketData.fromMap(data));
    } else if (opCode == 'hello') {
      return WebsocketPacket(OPCode.hello,
          data: HelloPacketData.fromJson(data));
    }
    throw 'Invalid packet $opCode';
  }
}

/// OPCodes for gateway
enum OPCode {
  // Command indicating request to gracefully close connection
  close,
  // Event indicating creation of a new message
  messageCreate,
  // Event indicating that a new user has joined
  userJoined,
  // Event indicating that a user has left
  userLeft,
  // Welcome packent containing channel data
  hello
}

/// Abstract data payload
abstract class WebSocketData {
  const WebSocketData();

  Map<String, dynamic> toJson();
}

/// Payload for new message
@JsonSerializable()
class MessageCreateData extends WebSocketData {
  /// The content of the message
  final String content;

  /// The name of the user creating the message
  final String author;

  /// Internal methods ///

  MessageCreateData(this.content, this.author);

  factory MessageCreateData.fromJson(Map<String, dynamic> map) =>
      _$MessageCreateDataFromJson(map);

  @override
  Map<String, dynamic> toJson() => _$MessageCreateDataToJson(this);
}

/// user join payload
@JsonSerializable()
class UserJoinedPacketData extends WebSocketData {
  /// The user which joined the channel
  final WebsocketUser user;

  /// Internal methods ///
  ///
  const UserJoinedPacketData(this.user);

  factory UserJoinedPacketData.fromMap(Map<String, dynamic> map) =>
      _$UserJoinedPacketDataFromJson(map);

  @override
  Map<String, dynamic> toJson() => _$UserJoinedPacketDataToJson(this);
}

/// user left payload
@JsonSerializable()
class UserLeftPacketData extends WebSocketData {
  /// The user which left the channel
  final WebsocketUser user;

  /// Internal methods ///
  const UserLeftPacketData(this.user);

  factory UserLeftPacketData.fromMap(Map<String, dynamic> map) =>
      _$UserLeftPacketDataFromJson(map);

  @override
  Map<String, dynamic> toJson() => _$UserLeftPacketDataToJson(this);
}

/// Representation of a client connected to a channel
@JsonSerializable()
class WebsocketUser {
  /// The name  of the client
  final String name;

  /// An unique identifier of the client
  final String nonce;

  /// Whether the client is an admin or not
  final bool isAdmin;

  /// Internal methods ///
  ///
  const WebsocketUser(this.name, this.nonce, this.isAdmin);

  factory WebsocketUser.fromJson(Map<String, dynamic> map) =>
      _$WebsocketUserFromJson(map);

  Map<String, dynamic> toJson() => _$WebsocketUserToJson(this);
}

/// Welcome packet to update client data
@JsonSerializable()
class HelloPacketData extends WebSocketData {
  /// The user of this client
  final WebsocketUser user;

  /// The key for this client
  final String key;

  /// The channelId the client connected to
  final String channelId;

  /// The name of the channel the client connected to
  final String channelName;

  /// A list of clients connected to the channel (including this client)
  final List<WebsocketUser> users;

  /// Internal methods ///

  const HelloPacketData(
      this.user, this.key, this.channelId, this.channelName, this.users);

  factory HelloPacketData.fromJson(Map<String, dynamic> map) =>
      _$HelloPacketDataFromJson(map);

  @override
  Map<String, dynamic> toJson() => _$HelloPacketDataToJson(this);
}
